### ISSUE TITLE
Add `-fstack-clash-protection` hardening (and enable by default?)

### ISSUE THREAD

**Issue by wizeman** [NixOS/nixpkgs#53753]
## Issue description

It would be good for the security of NixOS to compile user-space with `-fstack-clash-protection`. It should probably be added either as a new hardening flag or perhaps included as part of the `stackprotector` flag.

## Motivation

[CVE-2018-16864](https://www.qualys.com/2019/01/09/system-down/system-down.txt) and [CVE-2018-16865](https://www.qualys.com/2019/01/09/system-down/system-down.txt) describe new vulnerabilities and exploits in systemd which `-fstack-clash-protection` can mitigate, according to the linked advisory:

> SUSE Linux Enterprise 15, openSUSE Leap 15.0, and Fedora
28 and 29 are not exploitable because their user space is compiled with
GCC's -fstack-clash-protection

However, since NixOS doesn't compile user-space with `-fstack-clash-protection`, it's likely to be vulnerable.

**Comment by schmittlauch** [comment]
@wizeman Are you going to open an issue about actually patching systemd to fix the vulnerabilities, or shall I do it?

**Comment by wizeman** [comment]
@schmittlauch Please go ahead. I can't tell which systemd version has fixed those vulnerabilities from reading the advisory.

**Comment by globin** [comment]
@fpletz is actively working on this

**Comment by mkg20001** [comment]
@fpletz Any updates?

**Comment by Et7f3** [comment]
Do you see otherspot than [this](https://github.com/NixOS/nixpkgs/blob/f7ecc5425cf4047066a4e9315ee15979156ad845/pkgs/build-support/cc-wrapper/add-hardening.sh#L47) that need to be updated ?

**Comment by wizeman** [comment]
> Do you see otherspot than [this](https://github.com/NixOS/nixpkgs/blob/f7ecc5425cf4047066a4e9315ee15979156ad845/pkgs/build-support/cc-wrapper/add-hardening.sh#L47) that need to be updated ?

I would imagine there could be some fallout from adding the flag, i.e., it may break some packages, which would need to be fixed, updated or have the flag be removed from it.

Assuming my initial comment is still accurate as of today, then SUSE and Fedora have been using this flag for years already, so presumably the vast majority of packages should keep working fine.

**Comment by aikooo7** [comment]
Is this still needing to be fixed?

**Comment by wizeman** [comment]
> Is this still needing to be fixed?

Yes.

**Comment by tomodachi94** [comment]
Removing the 23.11 milestone. Pinging 24.11 release manager @RossComputerGuy, since I'm hesitant to add this to the 24.11 milestone.

**Comment by RossComputerGuy** [comment]
Yeah, I'm a little hesitant since we're passed the first point of blocking updates so we might be too far in the release cycle right now. However, I think we can tackle this post release.

**Comment by RossComputerGuy** [comment]
Was told to talk to @risicle. Any idea if what sort of build failure impact we could see if we do make this the default?

**Comment by risicle** [comment]
I _think_ the impact is probably quite minimal among core packages. I'm guessing this from the fact that we disable it for `wine`, which implies that people have built `wine` on a stdenv with stack clash protection enabled (perhaps via `pkgsExtraHardening`), and `wine` has an absurd dependency footprint. So I would guess those all build successfully with protection enabled.

As for leaf-y packages, I can't say - it's generally difficult to say anything about that until we just try it.

Though I will say that I tend not to introduce a hardening flag and enable it by default in one NixOS release, so I'd only suggest enabling it by default after november.
